### DISCUSSION TITLE
Add ignore,validation

### DISCUSSION THREAD

**Discussion by ujjwalguptaofficial** [ujjwalguptaofficial/JsStore#212] [General]
* ignore - https://jsstore.net/tutorial/insert/#ignore
* validation - https://jsstore.net/tutorial/insert/#validation

<hr /><em>This discussion was created from the release <a href='https://github.com/ujjwalguptaofficial/JsStore/releases/tag/4.0.2'>Add ignore,validation</a>.</em>

**Comment by a-l-e-c** [comment]
Just gut stuck for a few hours trying to figure out why autoIncrement value was not getting set in the meta.

Scenario:
I created a procedure to export all my data to `json` files and also a way to import the data back in again using those same files. 

The schema of the table in question was:
```js
var tblDataTracker = {
    name: 'DataTracker',
    columns: {
        Id:{ primaryKey: true, autoIncrement: true },
        dataId:  { notNull: false, dataType: JsStore.DATA_TYPE.Number },
        totalUpdated:  { notNull: false, dataType: JsStore.DATA_TYPE.Number, default: 0 },
        isComplete:  { notNull: false, dataType: JsStore.DATA_TYPE.Number, default: 0 },
        startDate: { notNull: true, dataType: JsStore.DATA_TYPE.DateTime },
    }
};
```

After clearing the table or using a new db, I was unable to import the data unless I use `validation: false`.
This might be something you wish to mention in the documentation.
Also, using `skipDataCheck: true` will solve the error for "date_time" fields, but not the autoIncrement.

This is my import function:
```js
async function importFromJson(forceColumnMatching=false, forceUpsert=false){
    importFormElement_Name = document.querySelector("#tblImportName");
    importFormElement_File = document.querySelector("#tblImportFile");
    tableName = importFormElement_Name.value;
    importFile = importFormElement_File;
    let file = importFile.files[0];
    let reader = new FileReader();
    reader.readAsText(file);
    reader.onload = async function() {
        let importObject = JSON.parse(reader.result);
        objectStoreNamesSet = new Set(jsstore_connection.database.tables);
        tableArray = Array.from(objectStoreNamesSet);
        tableObject = tableArray.filter(tbl => tbl.name == tableName);
        tableProperties = Object.values(tableObject)[0];
        importObjectKeys = Object.keys(importObject[0]);
        if(forceColumnMatching){
            //Compare incoming columns with existing table columns (ie. remove any data not matching table schema)
            importObject.forEach(item => {
                importObjectKeys.forEach((column, index) => {
                    if (!(column in tableProperties.columns)) {
                        delete item[column];
                    }
                });
            });
        }
        let upsertData = await jsstore_connection.insert({
            into: tableName,
            upsert: forceUpsert, //Update if entry already exists (requires primary key to be sent in values object)
            values: importObject
            ,validation: false //required when importing data that has autoincrement and/or date_time columns
            //,skipDataCheck : true //will prevent error on "date_time" fields but will not refresh autoIncrement value in JSStore meta
        }).then(function (newCount) {
            if(newCount >= 1){
                console.info(`SUCCESS ${tableName}.total:${newCount}`);
            }else{
                console.error(`ERROR ${tableName}.total:0`);
            }
        });
    };
    reader.onerror = function() {
        console.error(reader.error);
    };
}
```

**Reply by ujjwalguptaofficial** [reply]
@a-l-e-c - if i understand it correct : you are saying 

> After clearing table if you are inserting data again, autoIncrement is not working right ?

**Reply by a-l-e-c** [reply]
Not entirely. I'll try and explain better using detailed steps:
1. Clear the table (which has autoIncrement column)
2. Set `skipDataCheck: true` (to prevent error on any `date_time` columns) and recover data by importing from `.json` file
3. Confirm that data was recovered/imported into table.
4. Check `JsStore_Meta` table: the `autoIncColumnValue` of the table will still be `0`
5. Try now to add another field to the same table to see if the autoIncrement column works:
- Expected result: Field should be added with an incremented value on the autoIncrement column
- Instead, the following exception is thrown:
```
A mutation operation in the transaction failed because a constraint was not satisfied.
```

The only workaround is to use `validation: false` to avoid both the error on any `date_time` columns and also to prevent this autoIncrement problem.

**Reply by ujjwalguptaofficial** [reply]
its clearly mentioned @a-l-e-c 

```
# skipDataCheck

Do not check or change anything in data. By default value is false.

If supplied true, this will directly insert data without checking any thing like datatype, auto increment etc. This is useful in case - where you want to insert huge record at a time.
```

also diffrence b/w validation and skipDatacheck is there - 

```
Difference between validation & skipDataCheck is - validation generates autoIncrement field for autoIncrement column but skipDataCheck do not change anything in data.
```

when you imported - you need to make validation false, because you want to generate the autoincrement field or atleast sync it. 

Is there something missing ?

**Reply by a-l-e-c** [reply]
You are right, I got so carried away with trying to figure out what was going on instead of just going back to the documentation to read those important notes again :sweat_smile:

Would it be possible to implement a method for setting the value of autoIncrement after having done a large batch insert using `skidDataCheck`? This way people could take advantage of the speed of using skipDataCheck and still be able to continue using the table after the insert.

**Reply by ujjwalguptaofficial** [reply]
yeah that's what validation:false does. It does not validate others data but in case of autoincrement, it just keeps sync of . I have added further notes on the doc.

ofcourse it will be slower than skipDatacheck but in case of autoincrement and default option - you need to change data being inserted. 

I am pointing you to code for better clearity - https://github.com/ujjwalguptaofficial/JsStore/blob/ea85b5208ef394219201da5a5efa601a584ebf46/src/worker/executors/insert/values_checker.ts#L79

right now - i want to praise myself after looking at code :stuck_out_tongue: .

**Reply by a-l-e-c** [reply]
> ofcourse it will be slower than skipDatacheck but in case of autoincrement and default option - you need to change data being inserted.

The only workaround is to do:
_Note: This is only useful for importing to a browser which previously had the data_
1. Export data to `.json`
2. Use `remove` instead of `clear` on table:
```
var rowsDeleted = await jsstore_connection.remove({
    from: "largeTableName",
});
```
3. Using `skipDatacheck`: Import `.json` data (making sure to use `new Date()` for date_time columns)

But yes, it's still best (and safer!) to simply accept the fact that when dealing with autoIncrement columns, your import will be a slow(er) process on very large tables due to the fact that `validation` should be used and not `skipDatacheck`.

**Comment by ujjwalguptaofficial** [comment]
> Would it be possible to implement a method for setting the value of autoIncrement after having done a large batch insert using skidDataCheck

also regarding manual setup of autoincrement field - i would suggest not to do it. You might end up with something bad - let jsstore take care of it.
